### PR TITLE
Parser option to allow `new.target` outside functions

### PR DESCRIPTION
**Pull request by overlookmotel**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | n/a
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | n/a
| Any Dependency Changes?  | No
| License                  | MIT

This PR adds an option for `parse` to allow `new.target` outside of functions.

There are 2 places this can be useful:

1. In CommonJS code, `new.target` at top level is legal (if fairly pointless) as CommonJS code is compiled in a function wrapper.
2. When parsing code which will be wrapped in a function or execute in `eval()` where it inherits `new.target` from the outer scope.

I've called the option `allowNewTargetOutsideFunction` to match `allowAwaitOutsideFunction` and `allowReturnOutsideFunction`.

If accepted, I'll make a 2nd PR on the website repo to document this new option.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15114"></a>

### CODE REVIEW

**Review by liuxingbaoyu — APPROVED** [review]
LGTM, we generally need two approves and see what other people think.

**Comment by fisker** [comment]
I think this will make Prettier able to format [this](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgII7AA6UOOkUmMATgK5gwTUAUAlISWWRVTgGYQIOALw5OUOAHcAdDHTUA5nBgBuTjgC+JDSAA0ICAAcYAS2iZkoedQiSACvIQWU6VJPQBPC-oBG1dGAA1soAyugAtnAAMiYSyHyumHC+-kGhhgGxCsg0tMkgcOE+cAAmJaVR6FAKtOhKAGJM4egwptXIIOi0jHogABYw4agA6n0m8JgZYHAhTuMmAG7jHh1g2L2xSdQwdv4KzfGJ+QBWmAAeIVmocACKtBDwh6hJ+hnUWx0+6MWovYbUsRgwxMJRgfWQAA4AAyvGxJYb+Qwdf5wLYLOC9ACO93guyMzk6mAAtBJSqVetQ4NiTJTdnUDkgEs98klwiYcnQWVdbjiMYyjvo5D5gaDwUgAEyC-wmVBZADCEHCDIKmAArL1aEkACrfZxMl4gBZ5ACSUHKsBCYABxjwZpCMA81yeBv81Rmcm2yAAzAA2fSupQAUTNyFVqo0GiAA)

**Comment by fisker** [comment]
Maybe `allowNewTargetOutsideClass`?

**Comment by liuxingbaoyu** [comment]
@fisker 
I test this in the [babel repl](https://babeljs.io/repl#?browsers=&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1rAPYDsIBcBOArsDungBQCUiKqaWu0AZuutALzSYCmA7gHQ4weAOZccAbhoBfJFKA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=true&targets=&version=7.20.1&externalPlugins=&assumptions=%7B%7D) without errors, and also the prettier doesn't use recoverable exceptions?

**Comment by fisker** [comment]
That example is a "range format" feature. Only take few lines to format.

**Comment by liuxingbaoyu** [comment]
Personally it seems to me that it is normal to parse only part of it (eg pasting inside vscode), but this exception should be ignored by prettier.

**Review by nicolo-ribaudo — APPROVED** [review]
This seems ok. I would eventually like to replace all these `allow*` options with something like `context: "async function"`, or `context: "class method"`, etc.

We just released 7.20.0, so it will be a while before the 7.21.0 (which will include this PR).

@fisker You have to parse the whole file (and just print the specific section) regardless of this, otherwise you cannot properly format this example: https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0BUABKgBQCUhwAOlIYQIyEDUhJAnjnADYAmRAzGRoBfEABoQEAA65oAZ2SgAhgCcVEAO4AFVQgUolXDUrYKJAIxVKwAazgwAykoC2cADI4ocZKkNy4Fla29g5S1p4A5sgwKugBIHDO5nA8PCluSlAR6EoRcABiECrOSjC4WcggSugwEOIgABYwzlwA6g048HJhYHAOep04AG6dbJVgcmYgnv4qMFpWESU+fvEAVnIAHg6RXHAAiugQ8Ctc-hJhKrOV5krJXPVSKp4wrTg8MA3IABwADBfqfytKxSSpPOCzIbeCQARyO8AW0n0VTkAFovCkUvUVHA4TgcQtcsskL4zvF-M4cNFYuTdgd4d4SasJDA7m8Pl8kAAmFlWHBcSIAYQgzmJCTkAFZ6uh-AAVO76UnnEBDOIASSgaVgDjAzxkAEFNQ4YGw9qdlVYsn1WXNkPRvhJLXkAKKa5D8LnCYRAA

**Comment by fisker** [comment]
Good example, I'll try to fix that later.

**Comment by overlookmotel** [comment]
Thanks for the review. What's the process for submitting PR to update docs? Do it now, wait until this PR is merged, or wait until 7.21.0 is ready for release?

File: packages/babel-parser/src/options.ts
Comment (JLHwung) [inline packages/babel-parser/src/options.ts:46]: The current implementation also allows `new.target` within an arrow function. If it is desired, can you update the note and add a new test?

Comment (overlookmotel) [reply packages/babel-parser/src/options.ts:46]: To clarify: Are you saying that "top level" here is the wrong words to use because inside an arrow function is not top level?

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/options.ts:46]: Given that
```js
function fn() {
  () => new.target
} 
```
is valid, I would expect `() => new.target` to be valid with this option, right?

Comment (JLHwung) [reply packages/babel-parser/src/options.ts:46]: I don't think "top level" is wrong. But maybe it is incomplete in this case, since `() => new.target` will also be allowed.

Comment (overlookmotel) [reply packages/babel-parser/src/options.ts]: Yes, intent with this option was to allow `new.target` anywhere. If the file was wrapped in a function e.g. if it's a CommonJS file, I'm not aware of anywhere where `new.target` can't be used - because there's always a function above it.

Have added tests for `new.target` in an arrow function and updated the description of the option accordingly.

**Review by JLHwung — APPROVED** [review]

**Pull request merged**
